### PR TITLE
WIP: ubernetes federated service controller.

### PR DESCRIPTION
**Pull request by ghost**
Work in progress.  No review necessary yet.

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/24009)

### CODE REVIEW

File: federation/pkg/controller/service/servicecontroller.go [hunk 1]
```diff
@@ -0,0 +1,827 @@
...
+		}
+		return true
+	}
+}
+
+// clusterServiceSyncLoop handles updating the load balancers pointed to by all federated service DNS Records
+// whenever the set of cluster services in the federation changes.
+// TODO: quinton: Assumed that cluster service load balancers have host (i.e. DNS) names (and IP's?).  Verify and fix as necessary.
+//                Also the logic here bundles all hos/DNS names of all services in the federation together - need to handle service one separately.  Rewrite.
+func (s *ServiceController) clusterServiceSyncLoop(period time.Duration) {
```
Comment (mfanjie) [inline federation/pkg/controller/service/servicecontroller.go]: @quinton-hoole, about the last statement "if there is newly added cluster, add all services to it." in #23848 . My previous idea is use a informercontroller to watch the change of clusters, if there is new cluster being added, then add all services to it in dns server, and if there is a cluster being removed, then removed all entries related to this cluster. 

But this depends on how the dns provider api is designed. I found the current cloud provider api need all hosts to be passed in, if dns provider follow the same, then I do not expect a rewrite on this function... as we always need to pass all current available clusters into the api.
```
EnsureLoadBalancer(service *api.Service, hosts []string, annotations map[string]string) (*api.LoadBalancerStatus, error)
UpdateLoadBalancer(service *api.Service, hosts []string) error
```

File: federation/pkg/controller/service/servicecontroller.go [hunk 2]
```diff
@@ -0,0 +1,827 @@
...
+		cache.DeltaCompressorFunc(func(d cache.Deltas) cache.Deltas {
+			if len(d) == 0 {
+				return d
+			}
+			return cache.Deltas{*d.Newest()}
+		}),
+		s.cache,
+	)
+	lw := cache.NewListWatchFromClient(s.kubeClient.(*clientset.Clientset).CoreClient, "services", api.NamespaceAll, fields.Everything())
+	cache.NewReflector(lw, &api.Service{}, serviceQueue, serviceSyncPeriod).Run()
```
Comment (bprashanth) [inline federation/pkg/controller/service/servicecontroller.go]: Without doing any sort of review: We already know we should rewrite the existing serivice controller: https://github.com/kubernetes/kubernetes/issues/21625, so i'd suggest not doing things the way it does but instead following the patters in some of the other controllers in pkg/controller (eg: replication).

Comment (mfanjie) [inline federation/pkg/controller/service/servicecontroller.go]: @bprashanth Thanks for giving me more background. Will look into it and apply best patterns.

Comment (mfanjie) [inline federation/pkg/controller/service/servicecontroller.go]: I took a look on all related discussion for the rewriting of service controller, the discussion is regarding to how to watch the service deltas, and handle them in 10 go routines, and I agree that part need to be refactored to apply the best pattern like other controllers.

My comment here is not for the service delta, the logic is special for service controller, and there is no extra go routine to handle the logic. This function is watching the logic of k8s cluster changes, it obverse the changes and apply all services to all latest clusters. We can have more concrete discussions related to this part when dns api is ready.

Comment (bprashanth) [inline federation/pkg/controller/service/servicecontroller.go]: My comment was about using a reflector to watch + raw fifo, vs the informer work queue pattern. You probably don't want to manage the watch and fifo yourself, unless you really need to.

File: federation/pkg/controller/service/servicecontroller.go (hunk 1 above)
Comment (quinton-hoole) [inline federation/pkg/controller/service/servicecontroller.go]: @mfanjie Right now the dnsprovider interface accepts a single hostname, but I plan to add the multi-hostname function also, as you suggest.  The API will look almost identical to the above.

**Comment by quinton-hoole** [comment]
Replaced by  #26034 and related PR's

**Pull request closed without merging**